src/librustc_typeck/check/_match.rs

```diff
@@ -167,6 +170,25 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             prior_arm_ty = Some(arm_ty);
         }
 
+        // If all of the arms in the 'match' diverge,
+        // and we're dealing with an actual 'match' block
+        // (as opposed to a 'match' desugared from something else'),
+        // we can emit a better note. Rather than pointing
+        // at a diverging expression in an arbitrary arm,
+        // we can point at the entire 'match' expression
+        match (all_arms_diverge, match_src) {
+            (Diverges::Always { .. }, hir::MatchSource::Normal) => {
+                all_arms_diverge = Diverges::Always {
+                    span: expr.span,
+                    custom_note: Some(
+                        "any code following this `match` expression is unreachable, \
+                        as all arms diverge"
+                    )
+                };
+            },
+            _ => {}
+        }
+
         // We won't diverge unless the discriminant or all arms diverge.
         self.diverges.set(discrim_diverges | all_arms_diverge);
```

Comment:
Can this + the new bit above be refactored into a method? (I'd like to avoid too much non-spec logic in `fn check_match`).

Reply:
Such a method would need to take `all_arms_diverge`, `match_src`, `expr`, and `discrim_diverges` as parameters. I think that would make the code much harder to read, and would split the diverge logic over two methods (there are several other uses of `self.diverges` in this method).

Reply:
Well... it depends on *what* you intend to read. If you are writing text for the reference or auditing the reviewing of the language you don't care about `Diverges` at all because it only concerns a lint and so it is more readable that it not obstruct anything.

Rustfmt would display the following on a single line:

```rust
fn f1() {
    impl I {
        fn f2() {
            self.set_diverges_match(expr, match_src, discrim_diverges, all_arms_diverge);
        }
    }
}
```